tests/kernel/timer/timer_behavior/src/tick_timer_train.c

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright (c) 2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/sys/time_units.h>
+#include <zephyr/tc_util.h>
+#include <zephyr/ztest.h>
+
+#define TIMERS 1
+#define TEST_SECONDS 10
+#define MAX_CALLBACKS (CONFIG_SYS_CLOCK_TICKS_PER_SEC*TEST_SECONDS)
+
+struct timer_wrapper {
+	uint32_t callbacks;
+	struct k_timer tm;
+	uint32_t last_isr;
+	int64_t last_scheduled;
+	uint32_t max_delta;
+};
+
+K_SEM_DEFINE(timers_sem, 0, K_SEM_MAX_LIMIT);
+
+struct timer_wrapper timers[TIMERS];
+
+void tm_fn(struct k_timer *tm)
+{
+	struct timer_wrapper *tm_wrap =
+		CONTAINER_OF(tm, struct timer_wrapper, tm);
+	uint32_t now = k_cycle_get_32();
+
+	if (tm_wrap->last_isr != 0) {
+		uint32_t delta = now - tm_wrap->last_isr;
+
+		tm_wrap->max_delta = delta > tm_wrap->max_delta ? delta : tm_wrap->max_delta;
+	}
+	tm_wrap->last_isr = now;
+	tm_wrap->callbacks++;
+	if (tm_wrap->callbacks >= MAX_CALLBACKS) {
+		k_timer_stop(tm);
+		k_sem_give(&timers_sem);
+	} else {
+		int64_t next = tm_wrap->last_scheduled + TIMERS;
+
+		tm_wrap->last_scheduled = next;
+		k_timer_start(tm, K_TIMEOUT_ABS_TICKS(next), K_NO_WAIT);
+	}
+}
+
+
+/**
+ * @brief Test timers can be scheduled 1 tick apart without issues
+ *
+ * Schedules timers with absolute scheduling with a 1 tick
+ * period. Measures the total time elapsed and tries to run
+ * some fake busy work while doing so. If the print outs don't show up or
+ * the timer train is late to the station, the test fails.
+ */
+ZTEST(timer_behavior, test_one_tick_timer_train)
+{
+	const uint32_t max_time = TEST_SECONDS*1000 + 1000;
+
+	TC_PRINT("Initializing %u Timers, Tick Rate %uHz, Expecting %u callbacks in %u ms\n",
+		 TIMERS, CONFIG_SYS_CLOCK_TICKS_PER_SEC, MAX_CALLBACKS, max_time);
+
+	for (int i = 0; i < TIMERS; i++) {
+		k_timer_init(&timers[i].tm, tm_fn, NULL);
+		timers[i].max_delta = 0;
+	}
+
+	TC_PRINT("Starting Timers with Skews\n");
+	int64_t tick = k_uptime_ticks();
+
+	for (int i = 0; i < TIMERS; i++) {
+		timers[i].last_scheduled = tick+1;
+		k_timer_start(&timers[i].tm, K_TIMEOUT_ABS_TICKS(tick + 1), K_NO_WAIT);
+	}
+
+#ifdef CONFIG_TIMER_HAS_64BIT_CYCLE_COUNTER
+	uint64_t start_cycle = k_cycle_get_64();
+#else
+	uint32_t start_time_ms = k_uptime_get();
+#endif
+
+	uint32_t remaining_timers = TIMERS;
+
+	/* Do work in the meantime, proving there's enough time to do other things */
+	uint32_t busy_loops = 0;
+
+	while (true) {
+		while (k_sem_take(&timers_sem, K_NO_WAIT) == 0) {
+			remaining_timers--;
+
+		}
+		if (remaining_timers == 0) {
+			break;
+		}
+		TC_PRINT("Faking busy work, remaining timers is %u, timer callbacks %u\n",
+			 remaining_timers, timers[0].callbacks);
+		busy_loops++;
+		k_busy_wait(250000);
```

Comment:
This time looks wrong.  250ms is much, much longer than a tick, why wait so long when the semaphore gives should be arriving 10-100x faster?

Reply:
> Some notes. On the whole I think the test is a great idea, but I do caution that this is all but guaranteed to produce a flood of false positives on SMP qemu platforms in CI. Timer jitter due to host scheduling is a known (and largely unresolvable) thing, and this is a flashlight shining right on the problem.

Well the test case does try and exclude simulators already in testcase.yaml by tying it to the the existing timer behavior test. Though that has other issues. On arm the order of the tests (I need to see if I can change the ordering...) seems to be this timer train test followed by the behavior test, but by then the timer counter starts rolling over and the jitter_drift test fails because of that.